platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,815 @@
+3. Component Architecture
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+The public API changes consist of the following:
+
+* The Ledger Signature API
+* System Transactions
+* State Data Structures
+
+#### Ledger Signature API
+
+The Ledger Signature API is the interface between the application and the platform by which the application submits
+a message to be signed by the ledger, the nodes sign the message with their shares, and the platform aggregates the
+signatures into a ledger signature. The ledger signature is then asynchronously returned to the application.
+
+To request a ledger signature, the application invokes a new method on the `State`:
+
+* `State::sign(byte[] message) throws NotReadyException`
+
+To receive a ledger signature, the application registers a callback with the platform builder:
+
+TODO: Move signing API to TSS-Block-Signing proposal.
+
+* `PlatformBuilder::withLedgerSignatureCallback(Consumer<PairingSignature> ledgerSignatureConsumer)`
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
+platform. This value should not change during address book changes and its value does not change unless the network
+goes through a another TSS bootstrap process. Since its value is not expected to change, storing it in a singleton
+Merkle Leaf by itself is appropriate.
+
+TODO: Protobuf of data
+
+##### Consensus Rosters
+
+This proposal extends the roster data format where each `RosterEntry` has a new `elGamalKey` called a
+`tssEcKey` that is used in the Groth21 algorithm.
+
+TODO: Protobuf of data
+
+##### Key Material (TSS Data Map)
+
+The TSS Data Map is a combined map of all the key material for all rosters that are
+tracked and the votes by nodes that have validated the key material.
+
+TODO:  Separate the maps since the state API can handle them separate.
+
+Keys in the TSS Data Map have the following structure: (KeyType, RosterHash, SequenceNumber).
+
+1. KeyType - An enum that is one of TSS_MESSAGE or TSS_VOTE.
+2. RosterHash - The hash of the consensus roster that the message or vote is related to.
+3. SequenceNumber - The order in which the message or vote came to consensus.
+
+The value associated with a TSS_MESSAGE key is the pair (ShareId, TssMessage)
+
+1. ShareId - The id of the share that the message is related to.
+2. TssMessage - the raw bytes of the TssMessage
+
+TODO: protobuf for TssMessage (modify the TSS-Library proposal)
```

Comment:
I explained the need for shareId.   I agree.  I think the TssMessages should remain an opaque blob of bytes.    But I still need an absolute order on TssMessages so that I can use a bit-vector for voting.

Reply:
I think we reached consensus on our conversation.

Reply:
changing all usage to `tssEncryptionKey`